Faster SOURCE lookup with many entries

svenn-t opened the pull request:
With many entries in SOURCE, which can be the case when coupling other simulators to Flow, looking up a vector was very slow. Changing the structure to std::map makes looking up ijk indices much faster.

Removed the functions "hasHrate" and "hasTemperature" since "hrate" and "temperature" functions now return std::optional which can be used for checking instead.

Companion PR: https://github.com/OPM/opm-simulators/pull/6140

totto82 commented:
jenkins build this opm-simulators=6140 please

totto82 commented:
I never imagined a SOURCE for every cell in the initial implementation. Thanks for the improvement. The code changes are ok. I will merge this if Jenkins agrees and after I have done some local tests.

totto82 commented:
Local testing was done. I will merge this later today if there are no more comments from others.

bska commented on opm/input/eclipse/Schedule/Source.cpp:
```diff
@@ -76,154 +75,126 @@ Source::SourceCell::SourceCell(const DeckRecord& record) :
 
+    auto it = m_cells.find(ijk);
     if (it != m_cells.end()) {
-        *it = sourcenew;
+        auto it2 = std::find_if(it->second.begin(), it->second.end(),
+                               [&sourcenew](const auto& source)
+                               {
+                                   return source.isSame(sourcenew.component);
+                               });
+        if (it2 != it->second.end()) {
+            *it2 = sourcenew;
+        }
+        else {
+            it->second.emplace_back(sourcenew);
+        }
     } else {
-        this->m_cells.emplace_back(sourcenew);
+        m_cells.insert({ijk, {sourcenew}});
     }
```
Since `m_cells` now is a map, we can leverage the [`try_emplace()`](https://en.cppreference.com/w/cpp/container/map/try_emplace) member function, and especially its return value, to simplify this logic a little and avoid redundant element lookup operations:
```C++
auto [cellPos, inserted] = this->m_cells.try_emplace(ijk, { sourcenew });

if (! inserted) {
    auto sourcePos = std::find_if(cellPos->second.begin(), cellPos->second.end(),
                                  [comp = sourcenew.component](const auto& source)
                                  { return source.isSame(comp); });

    if (sourcePos != cellPos->second.end()) {
        *sourcePos = sourcenew;
    }
    else {
        cellPos->second.push_back(sourcenew);
    }
}
```

svenn-t replied:
Thanks for the suggestion! However, when I add the line:
```c++
auto [cellPos, inserted] = this->m_cells.try_emplace(ijk, { sourcenew });
```
I get compilation error:
```bash
Source.cpp:111:57: error: no matching function for call to ‘std::map<std::array<int, 3>, std::vector<Opm::Source::SourceCell> >::try_emplace(std::array<int, 3>&, <brace-enclosed initializer list>)’
```
I tried without the curly bracket around `sourcenew` but got a compilation error as well (something about "no matching function for call to ‘std::vector<Opm::Source::SourceCell>::vector(const Opm::Source::SourceCell&)").

bska replied:
> ```C++
> auto [cellPos, inserted] = this->m_cells.try_emplace(ijk, { sourcenew });
> ```

Okay, please try
```C++
auto [cellPos, inserted] = this->m_cells.try_emplace(ijk, std::vector { sourcenew });
```
instead.

bska replied:
The reason I'm a bit insistent on this is that I don't want the pattern of
```C++
if (find() != end()) {
    // do something
}
else {
    insert()
}
```
to become the way we interact with maps.

svenn-t replied:
The new suggestion worked, thanks a lot! It looks a bit better with `try_emplace`, so I understand the reasoning. 

I have updated the PR, where I have also have moved the `std::find_if` operations to a separate function (or rather two functions due to some `const` things I couldn't figure out).

bska replied:
> I have updated the PR, where I have also have moved the `std::find_if()` operations to a separate function (or rather two functions due to some `const` things I couldn't figure out).

Thanks&ndash;this looks better.  The `find_if()` helpers do strictly speaking not need to be member functions since they don't access any data members.  If you wanted to, you could just have a pair of free functions in the `.cpp` file instead.  You could even mash them together in a single template in an anonymous namespace in the `.cpp` file as
```C++
template <typename SourceCellSequence>
auto findInSequence(SourceCellSequence&& sequence, const Opm::SourceComponent comp)
{
    return std::find_if(std::begin(sequence), std::end(sequence),
                        [comp](const auto& source)
                        { return source.component == comp; });
}
```
Whether or not this is an improvement is, however, to some extent in the eye of the beholder.

svenn-t replied:
>  You could even mash them together in a single template in an anonymous namespace in the .cpp file as

Yes, this was exactly the how I wanted the code to work, but couldn't figure out for the life of me. Thanks! I have pushed an update.

bska reviewed (commented):
I think the `find_if()` calls should be factored out to a separately tested helper function to avoid having to write the correct implementation everywhere.  Other than that, there's one opportunity to improve the source insertion logic a little bit.

blattms commented:
How many entries do we expect in container m_cells (e.g. as percentage of total cells)?

svenn-t commented:
> How many entries do we expect in container m_cells (e.g. as percentage of total cells)?

In most cases, very few cells are used with SOURCE. But if you want to couple Flow with an external simulator through the SOURCE keyword, there can be one entry in m_cells per active cell.

totto82 commented:
jenkins build this opm-simulators=6140 please

svenn-t commented:
Jenkins build failed due to new SOURCE terms in the merged MICP pull request. I have rebased this PR and fixed the  https://github.com/OPM/opm-simulators/pull/6140 PR so it should build now.

bska commented:
jenkins build this opm-simulators=6140 please

totto82 commented:
Is this ready to be merged before the release branch deadline today?

bska reviewed (approved):
Thanks a lot for the updates.  This looks good to me now so I'll merge this and its downstream companion PR into the master branch.
